src/GraphQL/Execution/ExecutionStrategy.cs

```diff
@@ -357,15 +359,40 @@ void SetArrayItemNode(object d)
                         }
                         else if (node is ValueExecutionNode valueNode)
                         {
-                            node.Result = valueNode.GraphType.Serialize(d)
-                                ?? throw new InvalidOperationException($"Unable to serialize '{d}' to '{valueNode.GraphType.Name}' for list index {index}.");
+                            bool defaultException = false;
+                            try
+                            {
+                                object result = valueNode.GraphType.Serialize(d);
+                                if (result == null)
+                                {
+                                    defaultException = true;
+                                    if (listType.ResolvedType is NonNullGraphType)
+                                        throw new InvalidOperationException($"Unable to serialize '{d}' to '{valueNode.GraphType.Name}' for list index {index}.");
+                                    else
+                                        throw new ExecutionError($"Unable to serialize '{d}' to '{valueNode.GraphType.Name}'.");
+                                }
+                                node.Result = result;
+                            }
+                            catch (Exception error) when (listType.ResolvedType is NonNullGraphType)
+                            {
+                                // Stop execution as early as possible, it doesn't make sense to set the rest of the array elements and then execute nodes
+                                if (error is ExecutionError || defaultException)
+                                    throw;
+                                else
+                                    throw new InvalidOperationException($"Unable to serialize '{d}' to '{valueNode.GraphType.Name}' for list index {index}.", error);
```

Comment:
```suggestion
                                throw;
```
Exceptions are properly wrapped by the unhandled exception handler.  They should never be double-wrapped.  Perhaps it is too bad that they do not have the list index inside the message, but that's simply too bad.  Please remove.

Reply:
If there is an issue with the error's path property not returning the proper index for errors within non-null lists, then perhaps we need to solve this another way.  But we cannot double-wrap exceptions.

Reply:
Modify `Exception.Data` ?